internal/xds/translator/translator.go

```diff
@@ -120,12 +127,60 @@ func (t *Translator) Translate(ir *ir.Xds) (*types.ResourceVersionTable, error)
 	return tCtx, errs
 }
 
+// Maps XDS listeners to the extension policy instances that were registered to be sent to them
+type listenerHookPolicies map[string][]*ir.UnstructuredRef
+
+// notifyExtensionServerAboutListeners calls the extension server about all the translated listeners.
+func (t *Translator) notifyExtensionServerAboutListeners(
+	tCtx *types.ResourceVersionTable,
+	listenerExtPolicies listenerHookPolicies,
+) error {
+	// Return quickly if there is no extension manager or the Listener hook is not being used.
+	if t.ExtensionManager == nil {
+		return nil
+	}
+	if (*t.ExtensionManager).GetPostXDSHookClient(v1alpha1.XDSHTTPListener) == nil {
+		return nil
+	}
+
+	var errs error
+	for _, l := range tCtx.XdsResources[resourcev3.ListenerType] {
+		listener := l.(*listenerv3.Listener)
+		pols, found := listenerExtPolicies[listener.Name]
+		if !found {
+			pols = nil
+		} else {
+			slices.SortFunc(pols, func(l, r *ir.UnstructuredRef) int {
```

Comment:
this looks new, we are trying to ensure we pass in fixed order ?

Reply:
It's not about ensuring that we pass the policies in a fixed order. It's so that we don't duplicate them. The code sorts the list of extension policies and then compacts the list with `slices.CompactFunc` to remove duplicates.

See the [extensionpolicy-tcp-udp-http](https://github.com/envoyproxy/gateway/pull/3522/files#diff-07191b26ca09321c0d10a8cff082064e48c0729b420826c25ea781bb82f00f07) unit test for an example of how this can happen. The TL;DR is that when there are multiple listeners defined in the same gateway resource where the port is the same but one is for HTTP and the other is TCP, then in the translation to xDS the same xDS listener is reused because it's the same port. 

This code section is about making sure that if the same extension server policy is attached to different IR level constructs that map to the same xDS level construct, then the extension server won't receive the same policy twice.

Reply:
ah go it, thoughts on adding a `findIRListenesrByHostPort` method similar to 
https://github.com/envoyproxy/gateway/blob/e866ec138ca1666e55073d980f2d1aaf477f36d9/internal/xds/translator/translator.go#L598 which returns the IR listeners, allowing you to append/send the extensionRefs directly in here

this way we wont need to embed any extension policy code into the listener translator functions, mantain any state in memory, or perform any of these SortFuncs here

Reply:
Changed the logic to work like this. Not sure it's better, computation-wise.

Reply:
thanks, agree, its not better computation wise, but should improve readability and maintenance